src/lib/blkdata_svc/blkdata_service.cpp

```diff
@@ -53,87 +53,155 @@ void BlkDataService::create_vdev(uint64_t size, blk_allocator_type_t alloc_type,
+        return m_vdev->async_write(buf, size, blkid.to_single_blkid(), part_of_batch);
+    } else {
+        static thread_local std::vector< folly::Future< std::error_code > > s_futs;
+        s_futs.clear();
 
-    return m_vdev->alloc_blk(nblks, hints, out_blkids);
+        const char* ptr = buf;
+        auto blkid_it = blkid.iterate();
+        while (auto const bid = blkid_it.next()) {
+            uint32_t sz = bid->blk_count() * m_blk_size;
+            s_futs.emplace_back(m_vdev->async_write(ptr, sz, *bid, part_of_batch));
+            ptr += sz;
+        }
+        return collect_all_futures(s_futs);
+    }
 }
 
-void BlkDataService::commit_blk(const BlkId& bid) { m_vdev->commit_blk(bid); }
-
-blk_list_t BlkDataService::alloc_blks(uint32_t size) {
-    blk_alloc_hints hints; // default hints
-    std::vector< BlkId > out_blkids;
-    const auto status = alloc_blks(size, hints, out_blkids);
+folly::Future< std::error_code > BlkDataService::async_write(sisl::sg_list const& sgs, MultiBlkId const& blkid,
+                                                             bool part_of_batch) {
+    if (blkid.num_pieces() == 1) {
+        // Shortcut to most common case
+        return m_vdev->async_writev(sgs.iovs.data(), sgs.iovs.size(), blkid.to_single_blkid(), part_of_batch);
+    } else {
+        static thread_local std::vector< folly::Future< std::error_code > > s_futs;
+        s_futs.clear();
+        sisl::sg_iterator sg_it{sgs.iovs};
 
-    blk_list_t blk_list;
-    if (status != BlkAllocStatus::SUCCESS) {
-        LOGERROR("Resouce unavailable!");
-        return blk_list;
+        auto blkid_it = blkid.iterate();
+        while (auto const bid = blkid_it.next()) {
+            const auto iovs = sg_it.next_iovs(bid->blk_count() * m_blk_size);
+            s_futs.emplace_back(m_vdev->async_writev(iovs.data(), iovs.size(), *bid, part_of_batch));
+        }
+        return collect_all_futures(s_futs);
     }
+}
 
-    // convert BlkId to blklist;
-    for (auto i = 0ul; i < out_blkids.size(); ++i) {
-        blk_list.emplace_back(out_blkids[i].to_integer());
-    }
+BlkAllocStatus BlkDataService::alloc_blks(uint32_t size, const blk_alloc_hints& hints, MultiBlkId& out_blkids) {
+    HS_DBG_ASSERT_EQ(size % m_blk_size, 0, "Non aligned size requested");
+    blk_count_t nblks = static_cast< blk_count_t >(size / m_blk_size);
 
-    return blk_list;
+    return m_vdev->alloc_blks(nblks, hints, out_blkids);
 }
 
-folly::Future< bool > BlkDataService::async_free_blk(const BlkId bid) {
+void BlkDataService::commit_blk(MultiBlkId const& blkid) {
+    if (blkid.num_pieces() == 1) {
+        // Shortcut to most common case
+        m_vdev->commit_blk(blkid);
+    } else {
+        auto it = blkid.iterate();
+        while (auto const bid = it.next()) {
+            m_vdev->commit_blk(*bid);
+        }
+    }
+}
+
+folly::Future< std::error_code > BlkDataService::async_free_blk(MultiBlkId const& bids) {
+    // Shortcut to most common case
```

Comment:
remove this comment as vdev::free_blk() did the heavy lift

Reply:
Ok